src/Orleans.CodeGenerator/Generators/SerializerGenerator.cs

```diff
@@ -495,46 +495,63 @@ private static List<FieldInfoMember> GetFields(WellKnownTypes wellKnownTypes, Se
                 }
             }
 
-            // Some reference assemblies are compiled without private fields.
-            // Warn the user if they are inheriting from a type in one of these assemblies using a heuristic:
-            // If the type inherits from a type in a reference assembly and there are no fields declared on those
-            // base types, emit a warning.
-            var hasReferenceAssemblyBase = false;
-            var referenceAssemblyHasFields = false;
-            var baseType = type.BaseType;
-            while (baseType != null &&
-                   !baseType.Equals(wellKnownTypes.Object) &&
-                   !baseType.Equals(wellKnownTypes.Attribute))
-            {
-                if (!hasReferenceAssemblyBase && baseType.ContainingAssembly.HasAttribute("ReferenceAssemblyAttribute")) hasReferenceAssemblyBase = true;
-                foreach (var field in baseType.GetDeclaredMembers<IFieldSymbol>())
+            if (type.TypeKind == TypeKind.Class)
+            {
+                // Some reference assemblies are compiled without private fields.
+                // Warn the user if they are inheriting from a type in one of these assemblies using a heuristic:
+                // If the type inherits from a type in a reference assembly and there are no fields declared on those
+                // base types, emit a warning.
+                var hasUnsupportedRefAsmBase = false;
+                var referenceAssemblyHasFields = false;
+                var baseType = type.BaseType;
+                while (baseType != null &&
+                       !baseType.Equals(wellKnownTypes.Object) &&
+                       !baseType.Equals(wellKnownTypes.Attribute))
                 {
-                    if (hasReferenceAssemblyBase) referenceAssemblyHasFields = true;
-                    if (ShouldSerializeField(wellKnownTypes, field))
+                    if (!hasUnsupportedRefAsmBase && baseType.ContainingAssembly.HasAttribute("ReferenceAssemblyAttribute") && !IsSupportedRefAsmType(baseType)) hasUnsupportedRefAsmBase = true;
```

Comment:
This line is very long and it's not easy to see that `hasUnsupportedRefAsmBase` gets mutated at the end of it. Better split into at least two lines for readability.